File: core/logger.py
"""Журнал сэмплов и событий."""

from __future__ import annotations

import threading
from collections import deque

from core.sample import GamepadSample, RawInputEvent

MAX_SAMPLES = 50_000
MAX_EVENTS = 100_000
RING_SIZE = 500


class EventLogger:
    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._samples: deque[GamepadSample] = deque(maxlen=MAX_SAMPLES)
        self._events: deque[RawInputEvent] = deque(maxlen=MAX_EVENTS)
        self._ring_samples: deque[GamepadSample] = deque(maxlen=RING_SIZE)
        self._ring_events: deque[RawInputEvent] = deque(maxlen=RING_SIZE)
        self._active = False
        self._test_id = ""

    @property
    def is_active(self) -> bool:
        with self._lock:
            return self._active

    def start(self, test_id: str = "") -> None:
        with self._lock:
            self._active = True
            self._test_id = test_id
            self._samples.clear()
            self._events.clear()

    def stop(self) -> None:
        with self._lock:
            self._active = False
            self._test_id = ""

    def add_sample(self, sample: GamepadSample) -> None:
        with self._lock:
            self._ring_samples.append(sample)
            if self._active:
                self._samples.append(sample)

    def add_event(self, event: RawInputEvent) -> None:
        with self._lock:
            self._ring_events.append(event)
            if self._active:
                if event.test_id == "" and self._test_id:
                    event.test_id = self._test_id
                self._events.append(event)

    def get_samples(self, since_ns: int = 0) -> list[GamepadSample]:
        with self._lock:
            if since_ns <= 0:
                return list(self._samples)
            return [s for s in self._samples if s.timestamp_ns >= since_ns]

    def get_events(self, since_ns: int = 0) -> list[RawInputEvent]:
        with self._lock:
            if since_ns <= 0:
                return list(self._events)
            return [e for e in self._events if e.timestamp_ns >= since_ns]

    def recent_samples(self, limit: int = RING_SIZE) -> list[GamepadSample]:
        with self._lock:
            items = list(self._ring_samples)
            return items[-limit:]

    def recent_events(self, limit: int = RING_SIZE) -> list[RawInputEvent]:
        with self._lock:
            items = list(self._ring_events)
            return items[-limit:]

    def clear_session(self) -> None:
        with self._lock:
            self._samples.clear()
            self._events.clear()

File: src-tauri/src/logger.rs
use std::collections::VecDeque;
use std::sync::{Arc, Mutex};

use crate::sample::{GamepadSample, RawInputEvent};

const MAX_SAMPLES: usize = 50_000;
const MAX_EVENTS: usize = 100_000;
const RING_SIZE: usize = 500;

#[derive(Clone, Default)]
pub struct EventLogger {
    inner: Arc<Mutex<Inner>>,
}

struct Inner {
    samples: VecDeque<GamepadSample>,
    events: VecDeque<RawInputEvent>,
    ring_samples: VecDeque<GamepadSample>,
    ring_events: VecDeque<RawInputEvent>,
    active: bool,
    test_id: String,
}

impl Default for Inner {
    fn default() -> Self {
        Self {
            samples: VecDeque::with_capacity(1024),
            events: VecDeque::with_capacity(1024),
            ring_samples: VecDeque::with_capacity(RING_SIZE),
            ring_events: VecDeque::with_capacity(RING_SIZE),
            active: false,
            test_id: String::new(),
        }
    }
}

impl EventLogger {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn is_active(&self) -> bool {
        self.inner.lock().unwrap().active
    }

    pub fn start(&self, test_id: &str) {
        let mut inner = self.inner.lock().unwrap();
        inner.active = true;
        inner.test_id = test_id.to_string();
        inner.samples.clear();
        inner.events.clear();
    }

    pub fn stop(&self) {
        let mut inner = self.inner.lock().unwrap();
        inner.active = false;
        inner.test_id.clear();
    }

    pub fn add_sample(&self, sample: GamepadSample) {
        let mut inner = self.inner.lock().unwrap();
        if inner.ring_samples.len() == RING_SIZE {
            inner.ring_samples.pop_front();
        }
        inner.ring_samples.push_back(sample.clone());
        if inner.active {
            if inner.samples.len() == MAX_SAMPLES {
                inner.samples.pop_front();
            }
            inner.samples.push_back(sample);
        }
    }

    pub fn add_event(&self, mut event: RawInputEvent) {
        let mut inner = self.inner.lock().unwrap();
        if inner.ring_events.len() == RING_SIZE {
            inner.ring_events.pop_front();
        }
        inner.ring_events.push_back(event.clone());
        if inner.active {
            if event.test_id.is_empty() && !inner.test_id.is_empty() {
                event.test_id = inner.test_id.clone();
            }
            if inner.events.len() == MAX_EVENTS {
                inner.events.pop_front();
            }
            inner.events.push_back(event);
        }
    }

    pub fn get_samples(&self) -> Vec<GamepadSample> {
        self.inner.lock().unwrap().samples.iter().cloned().collect()
    }

    pub fn get_events(&self) -> Vec<RawInputEvent> {
        self.inner.lock().unwrap().events.iter().cloned().collect()
    }

    pub fn recent_events(&self, limit: usize) -> Vec<RawInputEvent> {
        let inner = self.inner.lock().unwrap();
        inner
            .ring_events
            .iter()
            .rev()
            .take(limit)
            .cloned()
            .collect::<Vec<_>>()
            .into_iter()
            .rev()
            .collect()
    }
}
